shards/shards_test.go

```diff
@@ -958,6 +958,167 @@ func TestAtomCountScore(t *testing.T) {
+	}{
+		{
+			name: "ignore content from files that are in filetombs",
+			shards: []shard{
+				{
+					documents:      []simpleDocument{fooForbidden, barAtMain},
+					fileTombstones: []string{"foo.go"},
+				},
+			},
+			expectedDocuments: []simpleDocument{barAtMain},
+		},
+		{
+			name: "include results from other shards that haven't tombstoned the file",
+			shards: []shard{
+				{
+					documents:      []simpleDocument{fooForbidden, barAtMain},
+					fileTombstones: []string{"foo.go"},
+				},
+				{
+					documents: []simpleDocument{fooAtMain},
+				},
+			},
+			expectedDocuments: []simpleDocument{fooAtMain, barAtMain},
+		},
+		{
+			name: "filetombstones affect all branches in a shard",
+			shards: []shard{
+				{
+					documents: []simpleDocument{
+						fooForbidden,
+
+						// the contents from this file should all ignored even though it's using a different branch
+						fooAtRelease,
+
+						barAtMainAndRelease,
+					},
+					fileTombstones: []string{"foo.go"},
+				},
+			},
+			expectedDocuments: []simpleDocument{barAtMainAndRelease},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// setup
+
+			shardMap := make(map[string]zoekt.Searcher)
+			for i, shard := range test.shards {
+				repository := zoekt.Repository{Name: "repository"}
+
+				// set of unique branches in the repository
+				branchesSet := make(map[string]struct{})
+
+				var zoektDocuments []zoekt.Document
+
+				for _, d := range shard.documents {
+					// de-duplicate branches from given documents
+					for _, b := range d.Branches {
+						branchesSet[b] = struct{}{}
+					}
+
+					// lift simple documents into zoekt.Documents
+					zoektDocuments = append(zoektDocuments, zoekt.Document{Name: d.Name, Branches: d.Branches, Content: d.Content})
+				}
+
+				for b := range branchesSet {
+					repository.Branches = append(repository.Branches, zoekt.RepositoryBranch{Name: b})
+				}
+
+				if len(shard.fileTombstones) > 0 {
+					repository.FileTombstones = make(map[string]struct{})
+					for _, s := range shard.fileTombstones {
+						repository.FileTombstones[s] = struct{}{}
+					}
+				}
+
+				shardMap[strconv.Itoa(i)] = searcherForTest(t, testIndexBuilder(t, &repository, zoektDocuments...))
+			}
+			ss := newShardedSearcher(4)
+			ss.replace(shardMap)
+
+			opts := &zoekt.SearchOptions{
+				Whole:                  true,
+				EvaluateFileTombstones: true,
+			}
+
+			q := &query.Substring{Pattern: "common"}
+
+			// test
+			result, err := ss.Search(context.Background(), q, opts)
+			if err != nil {
+				t.Fatalf("Search(%v, %s): %v", q, opts, err)
+			}
+
+			// verify
+
+			// extract a subset of the fields from each received fileMatch for
+			// easier comparisons
+
+			var receivedDocuments []simpleDocument
+			for _, f := range result.Files {
+				receivedDocuments = append(receivedDocuments, simpleDocument{
+					Name:     f.FileName,
+					Branches: f.Branches,
+					Content:  f.Content,
```

Comment:
why include content? In fact, why include branches? Our implementation will exclude filenames no matter the branch. So we should just need to test on the set of filenames. This should also simplify the comparison code since you have is a `[]string` that you can sort (or just a `map[string]struct{}`)

Reply:
There is a test that reads from two shards and asserts the tombstoned foo.go doesn't show up in the results, but the results from the non-tombstoned foo.go in the other shard do. In this case, you can't assert on just file names - you have to look at the contents

Reply:
However, I think you are correct about the branches. Each file has unique content, so it's redundant to assert on the branches as well. I'll remove them from the comparison.